scripts/validate.sh

```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+golangci-lint run
+go mod tidy
+go mod verify
+if unclean=$(git status --porcelain --untracked-files=no); then
+  if [ -n "$unclean" ]; then
+    echo "Encountered dirty repo!";
+    echo "$unclean";
+    exit 1;
+  fi
+else
```

Comment:
You could use `set -e` at the top instead of adding an else condition for this

Reply:
Just tried this - I see the error code but the exit code is still 0. Not sure if I am doing something wrong, but I need the exit code to persist too so that the stage fails. 

For reference, I added `set -e` to the top of the file, then I change the `git status` command to `g status` (to produce a command not found error). I saw the error message, but `echo $?` reported a 0 exit code - which makes me think that this doesn't work as desired. Did I miss something?

Reply:
You could use the extra variable like you're currently doing
```
$ cat set-e.sh 
#!/bin/bash
set -e
unclean=$(false && echo 'hello')
if [ -n "$unclean" ] ; then
    echo 'Encountered dirty repo!'
    exit 1
fi
$ bash -x set-e.sh 
+ set -e
++ false
+ unclean=
$ echo $?
1
$ sed -i -e 's/false/true/' set-e.sh
$ bash -x set-e.sh 
+ set -e
++ true
++ echo hello
+ unclean=hello
+ '[' -n hello ']'
+ echo 'Encountered dirty repo!'
Encountered dirty repo!
+ exit 1
$ echo $?
1
```
Is the presence of `git` the specific worry? What about the presence of `go` or `golangci-lint`? It doesn't seem to be something we're concerned about in rancher/rancher's validate script. You could use `which git` to check before proceeding. `set -e` should still be used in any case to halt execution and return an error if any of the linters or `go mod` commands fail.

Reply:
> Is the presence of git the specific worry?

No that part isn't a big deal, I install all of the dependencies in-pipeline in before using the validate. When testing on my local, I had an issue with git where the command exited because it couldn't determine repo ownership or some such (was because I had the directory mounted as a volume). I realized that if this command failed for whatever reason, we would pass the CI. We've had issues like this before - and it created scenarios where people were able to merge in files without running go generate and the like which caused further issues for others down the road (can point you to such a scenario if you are curious). 

I thought I would take a shot at getting a more robust script together that protects us from such issues. I'll try what you've put above - thank you for elaborating.

Reply:
Change made, thank you for helping me with that.